// @aws-crypto/material-management/build/module/index.js
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export { AlgorithmSuiteIdentifier, AlgorithmSuite, CommitmentPolicy, CommitmentPolicySuites, SignaturePolicy, SignaturePolicySuites, MessageFormat, NonCommittingAlgorithmSuiteIdentifier, CommittingAlgorithmSuiteIdentifier, } from './algorithm_suites';
export { WebCryptoAlgorithmSuite } from './web_crypto_algorithms';
export { NodeAlgorithmSuite } from './node_algorithms';
export { Keyring, KeyringNode, KeyringWebCrypto } from './keyring';
export { KeyringTraceFlag } from './keyring_trace';
export { MultiKeyringNode, MultiKeyringWebCrypto } from './multi_keyring';
export * from './materials_manager';
export { NodeEncryptionMaterial, NodeDecryptionMaterial, } from './cryptographic_material';
export { isValidCryptoKey, isCryptoKey, keyUsageForMaterial, subtleFunctionForMaterial, } from './cryptographic_material';
export { WebCryptoEncryptionMaterial, WebCryptoDecryptionMaterial, } from './cryptographic_material';
export { isEncryptionMaterial, isDecryptionMaterial, } from './cryptographic_material';
export { unwrapDataKey, wrapWithKeyObjectIfSupported, } from './cryptographic_material';
export { decorateCryptographicMaterial, decorateWebCryptoMaterial, } from './cryptographic_material';
export { SignatureKey, VerificationKey } from './signature_key';
export { EncryptedDataKey } from './encrypted_data_key';
export { immutableBaseClass, immutableClass, frozenClass, readOnlyProperty, } from './immutable_class';
export { needs } from './needs';
export { cloneMaterial } from './clone_cryptographic_material';
export * from './types';
export var EndOfSupportWarning;
(function (EndOfSupportWarning) {
    EndOfSupportWarning["v1"] = "This major version (1.x) of the AWS Encryption SDK for JavaScript has reached End-of-Support.\nIt will no longer receive security updates or bug fixes.\nConsider updating to the latest version of the AWS Encryption SDK.";
})(EndOfSupportWarning || (EndOfSupportWarning = {}));
